File: cpp/198. House Robber.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

/*
dp[i] = max(nums[i] + dp[i+2], dp[i+1])
dp[n] = 0
dp[n-1] = nums[n-1]
ans = dp[0]
*/

class Solution {
public:
    // int rob(vector<int>& nums) {
    //     int n = nums.size();
    //     if(n==0) return 0;
    //     int dp[n+1]; dp[n] = 0, dp[n-1] = nums[n-1];
    //     for(int i = n - 2; i >= 0; i--) {
    //         dp[i] = max(nums[i] + dp[i+2], dp[i+1]);
    //     }
    //     return dp[0];
    // }
    int rob(vector<int>& nums) {
        // 1D Space of above
        // Time: O(n), Space: O(1)
        int n = nums.size();
        if(n==0) return 0;
        int prev = 0, cur = nums[n-1];
        for(int i = n - 2; i >= 0; i--) {
            int cur_tmp = cur;
            cur = max(nums[i] + prev, cur);
            prev = cur_tmp;
        }
        return cur;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; int ans;

    solution = Solution();
    nums = {1,2,3,1};
    ans = solution.rob(nums);
    assert(ans==4);

    solution = Solution();
    nums = {2,7,9,3,1};
    ans = solution.rob(nums);
    assert(ans==12);
}

File: python/198-house-robber.py
from typing import List
from functools import lru_cache

'''
https://medium.com/@desolution/從leetcode學演算法-28-dynamic-programming-6-8e5a81a8437d
dp[i]: max money rob from nums[0] to nums[i]
dp[i] = max(dp[i-1], dp[i-2] + nums[i])
dp[0] = nums[0]
dp[1] = max(nums[0], nums[1])
'''
class Solution:
    def rob(self, nums: List[int]) -> int:
        "Down-Top DP, Time: O(n), Space: O(n) can be implemented to O(1)"
        n = len(nums)
        if n == 0: return 0
        if n == 1: return nums[0]
        if n == 2: return max(nums[0], nums[1])
        dp = [0 for _ in range(n)]
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, n):
            dp[i] = max(dp[i-1], dp[i-2] + nums[i])
        return dp[n-1]
    # def rob(self, nums: List[int]) -> int:
    #     "Top-Down Memoization, Time: O(n), Space: O(n)"
    #     if len(nums) == 0: return 0
    #     @lru_cache(None)
    #     def helper(i):
    #         if i == 0: return nums[0]
    #         if i == 1: return max(nums[0], nums[1])
    #         return max(helper(i-1), helper(i-2) + nums[i])
    #     return helper(len(nums)-1)

nums = [1,2,3,1]
result = Solution().rob(nums)
assert result == 4

nums = [2,7,9,3,1]
result = Solution().rob(nums)
assert result == 12

nums = [2,1,1,2]
result = Solution().rob(nums)
assert result == 4

nums = []
result = Solution().rob(nums)
assert result == 0